### ISSUE TITLE
shifting a       negative signed value is undefined

### ISSUE THREAD

**Issue by texrg** [mruby/mruby#3443]
```
mruby/mrbgems/mruby-sprintf/src/sprintf.c:20:32: note: expanded from
      macro 'EXTENDSIGN'
#define EXTENDSIGN(n, l) (((~0 << (n)) >> (((n)*(l)) % BITSPERDIG)) & ~(...
                            ~~ ^
~/mruby/mrbgems/mruby-sprintf/src/sprintf.c:37:11: warning: shifting a
      negative signed value is undefined [-Wshift-negative-value]
    *t |= EXTENDSIGN(3, strlen(t));
          ^~~~~~~~~~~~~~~~~~~~~~~~
~/mruby/mrbgems/mruby-sprintf/src/sprintf.c:20:76: note: expanded from
      macro 'EXTENDSIGN'
  ...l) (((~0 << (n)) >> (((n)*(l)) % BITSPERDIG)) & ~(~0 << (n)))
                                                       ~~ ^
```

**Comment by matz** [comment]
Hmm, I don't see the problem.  What is your compiler? clang?

Besides that, changing `EXTENDSIGN(3, strlen(t))` to `EXTENDSIGN(3U, strlen(t))` make any difference?

**Comment by tsahara** [comment]
Clang on macOS can check this undefined behavior though I don't see warning messages when I compile mruby source code on my mac.

```
% cat a.c
int
main(int argc, char **argv)
{
	return ~0 << 1;
}

% clang a.c
a.c:4:12: warning: shifting a negative signed value is undefined
      [-Wshift-negative-value]
        return ~0 << 1;
               ~~ ^
1 warning generated.
```

As clang complains, *shifting* a signed value is undefined.  From C11 draft:

> The result of E1 << E2 is E1 left-shifted E2 bit positions; (...) If E1 has a signed type and nonnegative value, and E1 × 2^E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined.

**Comment by matz** [comment]
@tsahara Hmm, does that mean changing 0 to 0U silence the warning?

**Comment by tsahara** [comment]
Yes, it does.

```
% cat a.c
int
main(int argc, char **argv)
{
	return ~0U << 1;
}

% clang a.c
%
```

**Comment by matz** [comment]
It has been fixed since e3bc81e7 on 2016-07-01.  Am I missing something?
If I am, reopen the issue.
